src/server/node_manager.c

```diff
@@ -2689,6 +2690,326 @@ recv_wk_job_idle(int stream)
+	if (new_exec_vnode == NULL) {
+		snprintf(err_msg, err_msg_sz-1,
+			"new_exec_vnode calloc error");
+		goto delete_from_exec_vnode_exit;
+	}
+
+	new_exec_vnode[0] = '\0';
+	entry = 0;	/* exec_vnode entries */
+	paren = 0;
+	for (chunk = parse_plus_spec_r(exec_vnode, &last, &hasprn);
+		chunk != NULL;
+		chunk = parse_plus_spec_r(last, &last, &hasprn)) {
+		paren += hasprn;
+		if (parse_node_resc(chunk, &noden, &nelem, &pkvp) == 0) {
+			if ((vnodelist != NULL) &&	
+			      !in_string_list(noden, '+', vnodelist)) {
+
+				if (entry > 0) { /* there's something */
+						 /* put in previously */
+					strcat(new_exec_vnode, "+");
+				}
+
+				if ( ((hasprn > 0) && (paren > 0)) ||
+				     ((hasprn == 0) && (paren == 0)) ) {
+						 /* at the beginning */
+						 /* of chunk for current host */
+					if (!parend) {
+						strcat(new_exec_vnode, "(");
+						parend = 1;
+					}
+				}
+				if (!parend) {
+					strcat(new_exec_vnode, "(");
+					parend = 1;
+				}
+				strcat(new_exec_vnode, noden);
+				entry++;
+
+				for (j = 0; j < nelem; ++j) {
+					snprintf(buf, sizeof(buf)-1, ":%s=%s",
+									pkvp[j].kv_keyw,
+									pkvp[j].kv_val);
+					strcat(new_exec_vnode, buf);
+				}
+
+				if (paren == 0) { /* have all chunks for */
+						  /* current host */
+			
+					if (parend) {
+						strcat(new_exec_vnode, ")");
+						parend = 0;
+					}
+				}
+			} else {
+
+				if (hasprn < 0) {
+					/* matched ')' in chunk, so need to */
+					/* balance the parenthesis */
+					if (parend) {
+						strcat(new_exec_vnode, ")");
+						parend = 0;
+					}
+				}
+			}
+		} else {
+			snprintf(err_msg, err_msg_sz-1,
+					"parse_node_resc error");
+			goto delete_from_exec_vnode_exit;
+		}
+
+	}
+
+	entry = strlen(new_exec_vnode)-1;
+	if (new_exec_vnode[entry] == '+')
+		new_exec_vnode[entry] = '\0';
+
+	rc = 0;
+	free(exec_vnode);
+	return (new_exec_vnode);
+
+delete_from_exec_vnode_exit:
+	free(exec_vnode);
+	free(new_exec_vnode);
+	return (NULL);
+
+}
+
+/**
+ * @brief
+ *	This removes 'pjob' from vnodes managed by parent mom 'pmom'.
+ *	Also, if pjob's 'exec_vnode_deallocated' attribute set,
+ *	then remove entries in 'exec_vnode_deallocated' that match
+ *	the vnodes where 'pjob' has already been taken out.
+ *
+ * @param[in]	pmom - the parent mom who sent the request.
+ * @param[in]	pjob - job in question
+ *
+ * @return void
+ */
+static void
+deallocate_job(mominfo_t *pmom, job *pjob)
+{
+	int   rc;
+        pbsnode *pnode;
+	int	i;
+	int	totcpus = 0;
+	int	totcpus0 = 0;
+	char	*freed_vnode_list = NULL;
+	int	freed_sz = 0;
+	char	*new_exec_vnode = NULL;
+	attribute deallocated_attr;
+	char	*jobid;
+
+	if ((pmom == NULL) || (pjob == NULL)) {
+		return;
+	}
+
+	jobid = pjob->ji_qs.ji_jobid;
+	if (*jobid == '\0')
+		return;
+
+	for (i = 0; i < svr_totnodes; i++) {
+		pnode = pbsndlist[i];
+
+		if ((pnode != NULL) && !(pnode->nd_state & INUSE_DELETED)
+			&& (pnode->nd_moms != NULL) &&
+				(pnode->nd_moms[0] == pmom)) {
```

Comment:
I've fixed this in the latest code line. I'm now going through all the parent moms and looking for a match.